### PR TITLE
Add support for Python 3

### PR DESCRIPTION
**Pull request by jwiggins**
This is an update to #1. I've added the `strength` submodule, cleaned up some formatting, and added a few more tests.

Also, if you want to turn on TravisCI for this repo, there's a config file now that exercises the library on Python 2.7, 3.4, and 3.5.

### CODE REVIEW

File: py/expression.cpp [hunk 1]
```diff
@@ -141,12 +141,13 @@ Expression_mul( PyObject* first, PyObject* second )
...
-{
-    return BinaryInvoke<BinaryDiv, Expression>()( first, second );
-}
+#if PY_MAJOR_VERSION < 3
+    static PyObject*
+    Expression_div( PyObject* first, PyObject* second )
+    {
+        return BinaryInvoke<BinaryDiv, Expression>()( first, second );
+    }
+#endif
```
Comment (sccolbert) [inline py/expression.cpp]: Why doesn't `__div__` work on Python 3?

File: py/pythonhelpers.h [hunk 2]
```diff
@@ -10,16 +10,32 @@
...
+#define MOD_INIT_FUNC(name) PyMODINIT_FUNC init##name(void)
+#endif
 
 #ifndef Py_RETURN_NOTIMPLEMENTED
 #define Py_RETURN_NOTIMPLEMENTED \
     return Py_INCREF(Py_NotImplemented), Py_NotImplemented
 #endif
 
+#ifndef cmpfunc
+#define cmpfunc int
```
Comment (sccolbert) [inline py/pythonhelpers.h]: would `void*` be better here instead of `int`?

File: py/expression.cpp (hunk 1 above)
Comment (jwiggins) [inline py/expression.cpp]: 2.7: https://docs.python.org/2.7/c-api/typeobj.html#number-object-structures
3.4: https://docs.python.org/3.4/c-api/typeobj.html#number-object-structures

`nb_divide` is gone in Py3k. Should it be moved to `nb_true_divide`? `nb_floor_divide`? Both?

File: py/pythonhelpers.h [hunk 3]
```diff
@@ -688,6 +704,10 @@ class PyDictPtr : public PyObjectPtr {
 
 };
 
+/* TODO - is there a better way than this?? */
+#ifndef PyMethod_GET_CLASS
+#define PyMethod_GET_CLASS(m) (PyObject*)Py_TYPE(PyMethod_GET_SELF(m))
```
Comment (sccolbert) [inline py/pythonhelpers.h]: Why is this necessary?

File: py/symbolics.h [hunk 4]
```diff
@@ -61,8 +61,10 @@ struct BinaryInvoke
 			return Invk()( primary, reinterpret_cast<Variable*>( secondary ) );
 		if( PyFloat_Check( secondary ) )
 			return Invk()( primary, PyFloat_AS_DOUBLE( secondary ) );
-		if( PyInt_Check( secondary ) )
-			return Invk()( primary, double( PyInt_AS_LONG( secondary ) ) );
+#if PY_MAJOR_VERSION < 3
+			if( PyInt_Check( secondary ) )
```
Comment (sccolbert) [inline py/symbolics.h]: there's an extra level of indentation inside this ifdef

File: py/term.cpp [hunk 5]
```diff
@@ -117,12 +117,13 @@ Term_mul( PyObject* first, PyObject* second )
...
 }
 
-
+#if PY_MAJOR_VERSION < 3
 static PyObject*
 Term_div( PyObject* first, PyObject* second )
 {
 	return BinaryInvoke<BinaryDiv, Term>()( first, second );
 }
+#endif
```
Comment (sccolbert) [inline py/term.cpp]: Ditto with `__div__`. We should should be able to divide expressions and terms on Python 3. There should be a different slot to fill on the Py3 type struct.

File: py/variable.cpp [hunk 6]
```diff
@@ -20,18 +20,32 @@ static PyObject*
 Variable_new( PyTypeObject* type, PyObject* args, PyObject* kwargs )
 {
 	static const char *kwlist[] = { "name", "context", 0 };
-	PyObject* name;
 	PyObject* context = 0;
+
+#if PY_MAJOR_VERSION >= 3
+	const char* name;
```
Comment (sccolbert) [inline py/variable.cpp]: who owns the `char*` after parsing like this?

File: py/pythonhelpers.h (hunk 2 above)
Comment (jwiggins) [inline py/pythonhelpers.h]: Here's how it's defined in Python 2.7:
`typedef int (*cmpfunc)(PyObject *, PyObject *);`

File: py/variable.cpp [hunk 7]
```diff
@@ -57,30 +71,36 @@ Variable_dealloc( Variable* self )
...
 }
 
 
 static PyObject*
 Variable_setName( Variable* self, PyObject* pystr )
 {
+#if PY_MAJOR_VERSION >= 3
+	if( !PyUnicode_Check( pystr ) )
+		return py_expected_type_fail( pystr, "unicode" );
+	self->variable.setName( _PyUnicode_AsString( pystr ) );
```
Comment (sccolbert) [inline py/variable.cpp]: I think `PyUnicode_AsUTF8` should be used instead of the deprecated typedef:
https://github.com/python/cpython/blob/master/Include/unicodeobject.h#L1156

But i'm not sure when that was added. I'm also not worried about support early Py3 versions.

File: py/variable.cpp [hunk 8]
```diff
@@ -135,11 +155,13 @@ Variable_mul( PyObject* first, PyObject* second )
 }
 
 
+#if PY_MAJOR_VERSION < 3
 static PyObject*
 Variable_div( PyObject* first, PyObject* second )
 {
 	return BinaryInvoke<BinaryDiv, Variable>()( first, second );
 }
+#endif
```
Comment (sccolbert) [inline py/variable.cpp]: another `__div__` issue

File: py/pythonhelpers.h (hunk 2 above)
Comment (sccolbert) [inline py/pythonhelpers.h]: Right, it's a function pointer. So defining it as an int would be incorrect. It needs to at least have the size of a pointer.

File: py/pythonhelpers.h (hunk 3 above)
Comment (jwiggins) [inline py/pythonhelpers.h]: The function has been removed in Python 3. Technically, the method which uses it isn't used anywhere in the wrapper. This and the relevant method in `PyMethodPtr` can be removed, if you like.

File: py/expression.cpp (hunk 1 above)
Comment (sccolbert) [inline py/expression.cpp]: `nb_true_divide`. Floor divide is the `//` operator IIUC.

File: py/variable.cpp (hunk 6 above)
Comment (jwiggins) [inline py/variable.cpp]: It appears to be owned by Python: https://docs.python.org/3.4/c-api/arg.html#strings-and-buffers

File: py/expression.cpp (hunk 1 above)
Comment (jwiggins) [inline py/expression.cpp]: Done

File: py/pythonhelpers.h (hunk 2 above)
Comment (jwiggins) [inline py/pythonhelpers.h]: `void*` works for me

File: py/symbolics.h (hunk 4 above)
Comment (jwiggins) [inline py/symbolics.h]: Fixed.

File: py/pythonhelpers.h (hunk 3 above)
Comment (sccolbert) [inline py/pythonhelpers.h]: or just ifdef out the offending part of `PyMethodPtr`

Comment (jwiggins) [inline py/pythonhelpers.h]: Will do

File: py/variable.cpp (hunk 6 above)
Comment (sccolbert) [inline py/variable.cpp]: Cool. It doesn't seem like we need the `length` parsing version either. Is the `s|0:__new__` a sufficient parse string?

File: py/variable.cpp (hunk 7 above)
Comment (jwiggins) [inline py/variable.cpp]: Agreed

File: py/variable.cpp (hunk 6 above)
Comment (jwiggins) [inline py/variable.cpp]: That seems like the correct one...

Comment (sccolbert) [inline py/variable.cpp]: Do you mean `s` or `s#` is the correct one? I don't think we should support anything but unicode variable names on Py3. `s#` also allows null bytes mid-string, which may cause other problems down the road.

Comment (jwiggins) [inline py/variable.cpp]: `s` seems like the correct one. I'm switching to that.

**Comment by jwiggins** [comment]
So... Wanna enable Travis-CI for this repository?

**Comment by sccolbert** [comment]
Travis is enabled. I'll have to add secure token once we merge this PR.

File: py/util.h [hunk 9]
```diff
@@ -74,12 +82,21 @@ convert_to_strength( PyObject* value, double& out )
 inline bool
 convert_to_relational_op( PyObject* value, kiwi::RelationalOperator& out )
 {
+#if PY_MAJOR_VERSION >= 3
+    if( !PyUnicode_Check( value ) )
+    {
+        PythonHelpers::py_expected_type_fail( value, "unicode" );
+        return false;
+    }
+    std::string str( _PyUnicode_AsString( value ) );
```
Comment (jwiggins) [inline py/util.h]: Missed two of these... hold on.

**Comment by jwiggins** [comment]
Hello Travis!

**Comment by jwiggins** [comment]
Right...
```
py/constraint.cpp:293:1: error: invalid conversion from ‘void*’ to ‘cmpfunc {aka int (*)(_object*, _object*)}’ [-fpermissive]
```

So... `int*`?

**Comment by jwiggins** [comment]
Actually, the whole mess is unnecessary. It's been removed.

**Comment by sccolbert** [comment]
Cool. I'm afk for a bit. Lemme know if/when everything is ready, and I'll try to merge tonight.

**Comment by jwiggins** [comment]
That's all I got, and Travis passed. Ball's in your court.

**Comment by jwiggins** [comment]
Thanks to @dwillmer for getting this started!
